WKWebExtensionAPITabs.ExecuteScript test has a race condition and fails sometimes.

xeenon opened the pull request:
#### 204a09687fe7b2ab4242dff1cf4c86577e198eb2
<pre>
WKWebExtensionAPITabs.ExecuteScript test has a race condition and fails sometimes.
<a href="https://webkit.org/b/264754">https://webkit.org/b/264754</a>
<a href="https://rdar.apple.com/problem/118344217">rdar://problem/118344217</a>

Reviewed by Brian Weinstein.

Use the onUpdated or onCompleted events to know when the tab navigates so we can execute script.
Also properly fire the onUpdated event by using web view navigation delegate methods.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab initWithWindow:extensionController:]): Set navigationDelegate.
(-[TestWebExtensionTab changeWebViewIfNeededForURL:forExtensionContext:]): Ditto.
(-[TestWebExtensionTab webView:didStartProvisionalNavigation:]): Added.
(-[TestWebExtensionTab webView:didReceiveServerRedirectForProvisionalNavigation:]): Added.
(-[TestWebExtensionTab webView:didFailNavigation:withError:]): Added.
(-[TestWebExtensionTab webView:didCommitNavigation:]): Added.
(-[TestWebExtensionTab webView:didFinishNavigation:]): Added.
(-[TestWebExtensionTab setParentTab:forWebExtensionContext:completionHandler:]): Added cast to fix error.

Canonical link: <a href="https://commits.webkit.org/270658@main">https://commits.webkit.org/270658@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3571010eb677d52e1477ad35fa7fd064da760d52

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26034 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4646 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/27317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/28133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/23845 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/26357 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6411 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/2075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/28133 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26288 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/3526 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/27317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/28714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3146 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/27317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/28714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23762 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/27317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/28714 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/3175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/2075 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/4567 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/27317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/3635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3342 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/3496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein reviewed (approved):
It might be a bit more straightforward to use webNavigation APIs here for future changes like this.

xeenon committed:
WKWebExtensionAPITabs.ExecuteScript test has a race condition and fails sometimes.
https://webkit.org/b/264754
rdar://problem/118344217

Reviewed by Brian Weinstein.

Use the onUpdated or onCompleted events to know when the tab navigates so we can execute script.
Also properly fire the onUpdated event by using web view navigation delegate methods.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionTab initWithWindow:extensionController:]): Set navigationDelegate.
(-[TestWebExtensionTab changeWebViewIfNeededForURL:forExtensionContext:]): Ditto.
(-[TestWebExtensionTab webView:didStartProvisionalNavigation:]): Added.
(-[TestWebExtensionTab webView:didReceiveServerRedirectForProvisionalNavigation:]): Added.
(-[TestWebExtensionTab webView:didFailNavigation:withError:]): Added.
(-[TestWebExtensionTab webView:didCommitNavigation:]): Added.
(-[TestWebExtensionTab webView:didFinishNavigation:]): Added.
(-[TestWebExtensionTab setParentTab:forWebExtensionContext:completionHandler:]): Added cast to fix error.

Canonical link: https://commits.webkit.org/270658@main
